### PR TITLE
Derive Instances For Recursive Data Types

### PR DESCRIPTION
**Pull request by adamgfraser**

### CODE REVIEW

**Review by sideeffffect — APPROVED** [review]
Sweet :+1:

File: core-tests/shared/src/test/scala-3/zio/prelude/DerivationSpec.scala [hunk 1]
```diff
@@ -25,6 +32,10 @@ object DerivationSpec extends ZIOSpecDefault {
         test("sealed trait") {
           assertTrue(Color.Red === Color.Red) &&
           assertTrue(Color.Red !== Color.Green)
+        },
+        test("recursive") {
+          assertTrue(NonEmptyList.Cons(1, NonEmptyList.Single(2)) == NonEmptyList.Cons(1, NonEmptyList.Single(2))) &&
+          assertTrue(NonEmptyList.Cons(1, NonEmptyList.Single(2)) != NonEmptyList.Cons(1, NonEmptyList.Single(3)))
```
Comment (sideeffffect) [inline core-tests/shared/src/test/scala-3/zio/prelude/DerivationSpec.scala]:
```suggestion
          assertTrue(NonEmptyList.Cons(1, NonEmptyList.Single(2)) === NonEmptyList.Cons(1, NonEmptyList.Single(2))) &&
          assertTrue(NonEmptyList.Cons(1, NonEmptyList.Single(2)) !== NonEmptyList.Cons(1, NonEmptyList.Single(3)))
```

**Review by sideeffffect — CHANGES_REQUESTED** [review]
typo :smile_cat:

File: core-tests/shared/src/test/scala-3/zio/prelude/DerivationSpec.scala (hunk 1 above)
Comment (adamgfraser) [reply core-tests/shared/src/test/scala-3/zio/prelude/DerivationSpec.scala]: Oops!

**Review by sideeffffect — APPROVED** [review]
I wanted to write that "we're all human", but that I realized that you're clearly not a mere human with all the work you're doing, so I have to retract that :smile:

**Comment by adamgfraser** [comment]
Definitely human! 😃

**Pull request merged**